### ISSUE TITLE
Gleam package manager seems to default to oldest possible version instead of most recent

### ISSUE THREAD

**Issue by dpbaines** [gleam-lang/gleam#3482]
I've been trying to pick up Gleam, part of that was going through this tutorial for a simple web app: https://gleaming.dev/articles/building-your-first-gleam-web-app/ 

I got stuck on the second step which was to run the line:
`gleam add wisp lustre mist gleam_http gleam_erlang dot_env gleam_json`

This adds the listed dependencies to the gleam.toml, with the widest version requirement range possible: 
```
gleam_stdlib = ">= 0.34.0 and < 2.0.0"
wisp = ">= 0.1.0 and < 1.0.0"
lustre = ">= 4.3.0 and < 5.0.0"
mist = ">= 1.2.0 and < 2.0.0"
gleam_http = ">= 3.6.0 and < 4.0.0"
gleam_erlang = ">= 0.25.0 and < 1.0.0"
dot_env = ">= 1.1.0 and < 2.0.0"
gleam_json = ">= 2.0.0 and < 3.0.0"
```
It then fetches the dependencies and creates a manifest.toml with the solved version numbers. 

I noticed a problem when trying to use the wisp library, and it not having code in it apart from the main function. I figured out that the package manager had defaulted to using the wisp version 0.1.0, with the hello world. I manually changed the requirements to force wisp version 0.16.0 and that fixed my problem. 

Digging deeper, it looks like the package manager calls PubGrub to solve it's version requirements, and PubGrub is suggesting the lowest possible version for each requirement. I looked into this, here's what I think so far. 

On this line: https://github.com/gleam-lang/gleam/blob/main/compiler-core/src/dependency.rs#L244 we are calling `choose_package_with_fewest_versions`. The [function docs](https://docs.rs/pubgrub/latest/src/pubgrub/solver.rs.html#272) say that it will greedily choose the first allowed version returned by `list_available_versions`. `list_available_versions` seems to be returning functions in order of oldest to newest. This seems like where it's going wrong. My instinct is that if we can change the sort order in this function it will resolve this.

Alternatively, this may be limited to my machine/runtime. Let me know if this is helpful, I have reproduced this a couple of times already. 

Gleam Version: 1.3.2
OS: Manjaro Linux Kernel: 6.2.16-2-MANJARO

**Comment by lpil** [comment]
Nice investigative work. A patch and a test to verify the behaviour would be very much welcome!

**Comment by pmbanugo** [comment]
Upvoting on this issue. I just wasted time debugging why my app was failing to build. Installing multiple packages (e.g. `gleam add gleam_http gleam_json gleam_erlang mist wisp`) uses old versions, but installing one after another uses the latest dependency.

**Comment by lpil** [comment]
Hello! Be sure to include the version specifiers when adding lots of dependencies like that.

**Comment by dpbaines** [comment]
I'll take a crack at this when I have a sec.

**Comment by pmbanugo** [comment]
> Hello! Be sure to include the version specifiers when adding lots of dependencies like that.

I just did. I think today was my day of dealing with package version error :(

**Comment by dpbaines** [comment]
Yeah I guess my initial analysis was completely off haha, it was just a dependency conflict caused by wisp requiring a version of mist that was outside the range that was default selected by gleam add. 
wisp required mist < 3.0 and my gleam.toml required mist >= 3.0

I would argue then that there are two other issues going on here that could be addressed:

1. It actually was pretty difficult figuring out _where_ the dependency conflict was coming from, this is the error message I got: 
```An error occurred while determining what dependency packages and
versions should be downloaded.
The error from the version resolver library was:

Unable to find compatible versions for the version constraints in your
gleam.toml. The conflicting packages are:

- gleam_crypto
- app
- wisp
- gleam_json
- mist
```
I think this could be clearer, like show which constraints are being violated and how

2. I feel like the ranges that `gleam add` picks are weirdly inconsistent, ie for mist `gleam add` chose the range >= 3.0.0 (the latest version of mist) and < 4.0.0 (?). For wisp it picked >= 0.1.0 (earliest version of wisp) to 1.0.0 (almost latest). It seems it's just picking the next major version as the top bound, but the bottom bound seems somewhat arbitrary? I might investigate this to see if there's something wrong going on there.

**Comment by dpbaines** [comment]
Fair, it seems like this is the case in other CLI package managers too. I'm going to close this issue then, it looks like that blog post I was following also updated the initial problem command haha
